File: leetcode_py/Reverse Linked List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        if not head or not head.next:
            return head
        dummy = ListNode(0)
        dummy.next = head
        cur = head
        while cur.next:
            tmp = cur.next.next
            cur.next.next = dummy.next
            dummy.next = cur.next
            cur.next = tmp
        return dummy.next


class Solution:
    def helper(self, head):
        if not head.next:
            self.head = head
            return head
        tail = self.helper(head.next)
        tail.next = head
        head.next = None
        return head

    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        if not head or not head.next:
            return head
        self.head = None
        self.helper(head)
        return self.head



class Solution:
    def f(self,head):
        if head.next==None:
            self.first=head
            return head 
        else:
            tail=self.f(head.next)
            tail.next=head
            return head 
        
    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        if head ==None:
            return None
        self.first=None
        self.f(head).next=None
        return self.first
    
if __name__=="__main__":
    a=Solution()
    p=ListNode(1)
    q=ListNode(2)
    p.next=q
    print a.reverseList(p).val
#

class Solution:
    def f(self,head):
        if head.next==None:
            self.head=head
            return head
        else:
            tmp=self.f(head.next)
            head.next=None
            tmp.next=head
            return head
    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        if head==None or head.next==None:
            return head
        self.head=None
        self.f(head)
        return self.head


class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        if head ==None:
            return None
        pre,p=None,head
        while p:
            tmp=p.next
            p.next=pre
            pre=p
            p=tmp
        return pre

class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        if head ==None:
            return None
        pre,p=head,head.next
        while p:
            tmp=p
            p=p.next
            tmp.next=pre
            pre=tmp 
        head.next=None
        return pre

File: algorithms/cpp/reverseLinkedList/reverseLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
// recursive
class Solution {
    ListNode* newhead;
public:
    ListNode* helper(ListNode* head){
        if(head->next == NULL) {newhead=head; return head;}
        ListNode* tail = helper(head->next);
        tail->next = head;
        head->next = NULL;
        return head;
    }
    ListNode* reverseList(ListNode* head) {
        if(head == NULL) return NULL;
        helper(head);
        return newhead;
    }
};
// Source : https://leetcode.com/problems/reverse-linked-list/
// Author : Hao Chen
// Date   : 2015-06-09

/********************************************************************************** 
 * 
 * Reverse a singly linked list.
 * 
 * click to show more hints.
 * 
 * Hint:
 * A linked list can be reversed either iteratively or recursively. Could you implement both?
 * 
 *               
 **********************************************************************************/


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList_iteratively(ListNode* head) {
        ListNode *h=NULL, *p=NULL;
        while (head){
            p = head->next;
            head->next = h;
            h = head;
            head = p;
        }
        return h;
    }
    ListNode* reverseList_recursively(ListNode* head) {
        if (head==NULL || head->next==NULL) return head;
        ListNode *h = reverseList_recursively(head->next);
        head->next->next = head;
        head->next = NULL;
        return h;
        
    }
    ListNode* reverseList(ListNode* head) {
        return reverseList_iteratively(head);
        return reverseList_recursively(head);
    }
};
